File: rag_gemini_pdf.js
const fs = require('fs');
const pdf = require('pdf-parse');
const { GoogleGenerativeAI } = require('@google/generative-ai');

// Function to load and chunk PDF (unchanged)
async function loadPDF(pdfPath) {
    const dataBuffer = fs.readFileSync(pdfPath);
    const data = await pdf(dataBuffer);
    return data.text;
}

function chunkText(text) {
    return text.split('\n\n'); // Simple chunking by paragraphs
}

// Updated getEmbeddings to accept extractor
async function getEmbeddings(texts, extractor) {
    const embeddings = [];
    for (const text of texts) {
        const output = await extractor(text, { pooling: 'mean', normalize: true });
        embeddings.push(Array.from(output.data));
    }
    return embeddings;
}

// Updated ragQuery to accept extractor
async function ragQuery(model, vectorStore, question, extractor, k = 5) {
    const queryEmbedding = await getEmbeddings([question], extractor);
    const relevantTexts = vectorStore.similaritySearch(queryEmbedding[0], k);
    const context = relevantTexts.join('\n\n');
    const prompt = `Answer the following question using the provided context.\n\nContext:\n${context}\n\nQuestion: ${question}`;
    const result = await model.generateContent(prompt);
    return result.response.text();
}

// Simple VectorStore class (unchanged)
class VectorStore {
    constructor() {
        this.vectors = [];
        this.texts = [];
    }
    add(vector, text) {
        this.vectors.push(vector);
        this.texts.push(text);
    }
    similaritySearch(queryVector, k) {
        // Placeholder for similarity search logic
        return this.texts.slice(0, k);
    }
}

function setupGemini(apiKey) {
    const genAI = new GoogleGenerativeAI(apiKey);
    return genAI.getGenerativeModel({ model: 'gemini-2.5-pro-exp-03-25' });
}

async function main() {
    const pdfPath = process.argv[2];
    const apiKey = process.argv[3];

    console.log(`🔍 Loading PDF: ${pdfPath}`);
    const text = await loadPDF(pdfPath);
    const chunks = chunkText(text);
    console.log(`✅ Chunked into ${chunks.length} segments.`);

    // Dynamically import the ES Module and create extractor
    const { pipeline } = await import('@xenova/transformers');
    const extractor = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');

    // Use extractor to generate embeddings
    const embeddings = await getEmbeddings(chunks, extractor);
    const vectorStore = new VectorStore();
    chunks.forEach((chunk, index) => {
        vectorStore.add(embeddings[index], chunk);
    });
    console.log("✅ Embeddings created and vector store built.");

    const model = setupGemini(apiKey);

    console.log("\n🤖 Ask questions about the PDF (type 'exit' to quit):");
    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });

    function askQuestion() {
        readline.question('>> ', async (query) => {
            if (query.toLowerCase() === 'exit' || query.toLowerCase() === 'quit') {
                readline.close();
                return;
            }
            const answer = await ragQuery(model, vectorStore, query, extractor);
            console.log(`\n📎 Answer:\n${answer}\n`);
            askQuestion();
        });
    }

    askQuestion();
}

main().catch(console.error);

File: rag_gemini_pdf.py
import os
import pdfplumber
import google.generativeai as genai
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.vectorstores import FAISS
from langchain.embeddings import SentenceTransformerEmbeddings
from langchain.docstore.document import Document

# 1. Load PDF
def load_pdf(path):
    full_text = ""
    with pdfplumber.open(path) as pdf:
        for page in pdf.pages:
            full_text += page.extract_text() + "\n"
    return full_text

# 2. Chunk PDF text
def chunk_text(text, chunk_size=1000, overlap=200):
    splitter = RecursiveCharacterTextSplitter(
        chunk_size=chunk_size,
        chunk_overlap=overlap
    )
    return splitter.create_documents([text])

# 3. Create vector database
def embed_documents(docs):
    embedding_model = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")
    vectorstore = FAISS.from_documents(docs, embedding_model)
    return vectorstore

# 4. Initialize Gemini
def setup_gemini(api_key):
    genai.configure(api_key=api_key)
    return genai.GenerativeModel("gemini-2.5-pro-exp-03-25")

# 5. Retrieval-Augmented QA
def rag_query(model, vectorstore, question, k=5):
    docs = vectorstore.similarity_search(question, k=k)
    context = "\n\n".join([doc.page_content for doc in docs])
    prompt = f"""Answer the following question using the provided context.

Context:
{context}

Question: {question}
"""
    response = model.generate_content(prompt)
    return response.text

# 6. Main CLI loop
def main(pdf_path, api_key):
    print(f"🔍 Loading PDF: {pdf_path}")
    text = load_pdf(pdf_path)
    docs = chunk_text(text)
    print(f"✅ Chunked into {len(docs)} segments.")

    vectorstore = embed_documents(docs)
    print("✅ Embeddings created and vector store built.")

    model = setup_gemini(api_key)

    print("\n🤖 Ask questions about the PDF (type 'exit' to quit):")
    while True:
        query = input(">> ")
        if query.lower() in {"exit", "quit"}:
            break
        answer = rag_query(model, vectorstore, query)
        print(f"\n📎 Answer:\n{answer}\n")

if __name__ == "__main__":
    import sys
    if len(sys.argv) != 3:
        print("Usage: python rag_gemini_pdf.py the_pdf.pdf MA_GEMINI_API_KEY")
    else:
        main(sys.argv[1], sys.argv[2])

